text/0000-array-private-classes

```diff
@@ -0,0 +1,105 @@
+- Feature Name: array_private_classes
+- Start Date: 2022-01-08
+- RFC PR: (leave this empty)
+- Pony Issue: (leave this empty)
+
+# Summary
+
+Using information hiding design principle, the builtin classes `ArrayKeys`, `ArrayValues` and `ArrayPairs` can be made private as they are only used as return types for `Array` functions `keys`, `values` and `pairs`. The return values for these functions is be changed to general `Iterator`. A new interface `Rewindable` is defined to allow for rewindable iterators.
+
+# Motivation
+
+This change brings:
+
+- Applying the design principle of [hiding implementation details](https://en.wikipedia.org/wiki/Information_hiding) but offer a general and stable interface. Returning interfaces instead of concrete classes allows changing the implementation.
+- `Array` functions `keys`, `values` and `pairs` definitions are made more general. `Array` iterators implementation details are not public. `_ArrayKeys`, `_ArrayValues` and `_ArrayPairs` are now [opaque data types](https://en.wikipedia.org/wiki/Opaque_data_type).
+- Makes the standard library more simple by removing 3 specialised classes.
+- The interface `Rewindable` can be combined to create rewindable iterators or other types that can be rewinded.
+
+This change is compatible with the existing code base but for client code that is directly using the classes `ArrayKeys`, `ArrayValues` and `ArrayPairs`. A search on Github shows that the impact is very limited.
+
+# Detailed design
+
+`Array` functions `keys`, `values` and `pairs` are changed to return `Iterator` and the classes that implement these iterators are make private. Here are the full implementation of these functions.
+
+```pony
+  fun keys(): Iterator[USize]^ =>
+    """
+    Return an iterator over the indices in the array.
+    """
+    _ArrayKeys[A, this->Array[A]](this)
+
+  fun values(): (Iterator[this->A] & Rewindable[this->A]) =>
```

Comment:
Rather than seeing `Rewindable` as a separate interface, I'd prefer to collapse the two together into an interface that is a subtype of `Iterator`, called `IteratorRewindable`:
```suggestion
  fun values(): IteratorRewindable[this->A] =>
```

My reasons are:
1. Most (all?) known use cases are for an iterator that is rewindable, not some non-iterator object that has a `rewind` method that returns an iterator
2. I think it's important to be really careful with the type names we put in `builtin`, because it means that no one else can ever declare a type with that name. I think `IteratorRewindable` is less likely to collide as compared to the other possible uses one can imagine for the name `Rewindable`.

In fact, one of the reasons I favor this change is because it would reduce the number of public type names defined in `builtin`, and thus reduce the collision surface for type names with those that users may want to define in their programs. So I personally see `IteratorRewindable` as being a better option for keeping that collision surface minimized, while also keeping it convenient to define an iterator that is rewindable (the suggested change reduces the total amount of characters to type/read, while keeping the intent very clear):

Reply:
In fact, the reason why I did not create a `RewindableIterator` interface but used `Rewindable` instead, is that this interface is not to dedicated to iterators. The interface defines a property that classes that implement `Rewindable` must implement a function named `rewind`. The semantic of that property is limited to the `rewind` name; it does not define a contract that `rewind` must implement, for instance that `has_next` must be true if it was true in a previous iteration and that `next` must give again the first element of the iterator.

Because of that, I thought that even if the risk of naming collision exists (one developer could already have a `Rewindable` type; the name is not used in Pony code on Github), the risk of semantic collision was null as `Rewindable` is an interface. But thinking while typing this, I now understand that the compiler will complain if there are two types with the same name but different definitions...

A `RewindableIterator` interface overlaps `Iterator` definition, something like:
```pony
interface RewindableIterator[A]
  fun has_next(): Bool
  fun next(): A ?
  fun ref rewind()
```

As Pony is using composition instead of inheritance to define new types, perhaps does it limit the ability to build new types with this interface? While one can add
```pony
type (Iterator[A] & Rewindable[A]) is RewindableIterator[A]
```
and get the same result with more flexibility...

On the other hand, while Pony's stdlib is still evolving, defining shorter builtin types and names while Pony's ecosystem is still developing is the right time to take such decisions because they can't be changed later.

In the end, as we both agree on naming of builtins is very important for future code, I can change `Rewindable` to whatever name will be decided. This only concern the type returned by the `values` function, as `keys` and `pairs` could also probably return such iterator type.

Reply:
> In fact, the reason why I did not create a `RewindableIterator` interface but used `Rewindable` instead, is that this interface is not to dedicated to iterators. The interface defines a property that classes that implement `Rewindable` must implement a function named `rewind`.

Well, your interface definition (as given in your document here) is not only that it must implement a function named `rewind`, but also that this function must return an `Iterator[A]`. Hence, your interface is only relevant to the use case of iterators and not to any arbitrary thing that calls for rewinding.

That's why I think it needs a more specific name - it is specifically about rewinding to produce an iterator - not about a fully abstract concept of rewinding anything.

Reply:
Remember that while there exist workarounds for preventing namespace collisions in libraries you import, the `builtin` library (the one where `Array` is defined) is special in that it is always present. This is why I place such emphasis on taking care not to pollute the `builtin` library with public type names that could collide with useful types that the user may want to define differently with that same name - `builtin` is the only library that will make defining such a type in user code literally impossible.

Reply:
Yes, you're right. I focused on the function name and I forgot about the return type.
Should the `RewindableIterator` interface define the 3 functions `has_next`, `next` and `rewind`, or only the `rewind` one to be used by composition? I think the former makes the code simpler by defining a new type of iterator.

Reply:
Yes, with a name like `RewindableIterator`, I'd expect it to be a subtype of `Iterator`, so I believe it should have all the same methods as `Iterator`, in addition to the `rewind` method.